async-iterator.md

```diff
@@ -0,0 +1,904 @@
+
+
+## Lowering
+
+I'll describe the process of **lowering** async-iterator methods, that is replacing awaits and yields with simpler primitives.
+
+I'll start with many techniques already used for async methods (await suspensions, extracting locals to fields, spilling awaits, dispatch blocks).
+Then I'll explain `yield return` suspensions, disposal and yield breaks.
+
+
+### `await`
+
+`await expr` in async-iterator methods is lowered as:
+```C#
+... code before `await expr` ...
+
+{
+    var awaiterTemp = <expr>.GetAwaiter();
+    if(!awaiterTemp.IsCompleted)
+    {
+        /* set state to N */
+        /* save awaiter temp */
+        /* kick off background execution */
+        return;
+
+        resumeLabelN:
+        /* reset state */
+        /* restore awaiter temp */
+    }
+}
+
+// continue method execution with the result
+... awaiterTemp.GetResult()
+
+... code after `await expr` ...
+```
+
+Note this is the same pattern used to lower awaits in async methods.
+
+As you can see, we're introducing a suspension and resumption in the middle of the method.
+The suspension is the code leading up to the `return` statement. Each suspension is identified by a number `N`.
+The resumption is the code starting from label `resumeLabelN:`.
+
+You might wonder how execution resumes from that label. The lowering also adds logic at the beginning of the method to jumps to `resumeLabelN:` when the **state** is `N`.
+So `{ ... method body... }` is lowered to start with a **dispatching** `switch` statement:
+```C#
+{
+    switch (state)
+    {
+        ...
+        case N:
+            goto resumeLabelN;
+        ...
+    }
+
+    ... rest of lowered method body, including the suspension for state N and the `resumeLabelN` label ...
+}
+```
+
+When either the caller or background execution need to move the method forward, execution will resume from the right label because the state variable was set to `N` and the dispatching logic will jump to that label.
+
+Because there is a cost to starting background execution, suspending and resuming, the lowering pattern is optimized: if the task was already completed, we avoid that overhead and instead just move ahead with the result.
+
+
+#### Locals
+
+Although I won't go into much details on this, it is worth mentioning that all the local variables in the method are converted to fields on a compiler-generated type, so they maintain their values across suspensions/resumptions.
+
+The compiler also generates fields for:
+- method parameters,
+- the `state` value,
+- persisting `awaiterTemp`,
+- the `current` value, (see section on "yield return" suspension)
+- the machinery for background execution and continuation.
```

Comment:
if these will get sections explaining them, consider having links here to those sections.  this feedback applies to this entire doc.

Reply:
stopping now.  it's too late for my brain to take on hte more complex stuff :)